test/utils/helpers.ts

```diff
@@ -1,26 +1,92 @@
 import { BigNumber } from 'ethers';
-import { ethers } from 'hardhat'  
+import { ethers } from 'hardhat';
+import { SignerWithAddress } from '@nomiclabs/hardhat-ethers/signers';
 
-
-export const zero = ethers.constants.Zero; 
+export const zero = ethers.constants.Zero;
 
 export interface Reward {
   _address: string;
   _amount: BigNumber;
+  _signer?: SignerWithAddress;
 }
 
 /**
- * @notice Sums rewards amounts that are due to voters 
- * @param rewards Array of structs of voters' addresses and amounts 
- * to be transfered 
+ * @notice Sums rewards amounts that are due to voters
+ * @param rewards Array of structs of voters' addresses and amounts
+ * to be transfered
  */
 
 export function sum(rewards: Reward[]): BigNumber {
-  let sum = zero;
+  let SUM = zero;
   for (let i in rewards) {
-    sum = sum.add(rewards[i]._amount);
+    SUM = SUM.add(rewards[i]._amount);
+  }
+  return SUM;
+}
+
+/**
+ * @notice Distributes a value randomly to a given array length such that
+ * the sum adds back to that initial value
+ * @param length array length which gets the value randomly distributed to
+ * @param value value we want to distribute randomly
+ */
+
+function distributeBN(length: number, value: number): BigNumber[] {
+  // internal function used for math computations in number type
+  function distribute(_length: number, _value: number): number[] {
+    if (_length <= 1) return [_value];
+    const half = Math.floor(_length / 2);
+    const dist = Math.floor(Math.random() * _value);
+    let result = distribute(half, dist).concat(distribute(_length - half, _value - dist));
+    return result;
+  }
+  let distributeArray = distribute(length, value);
+
+  // set up BN type array for function output
+  let distributeArrayBN: BigNumber[] = [];
+  // loop through each number in the array generated by the internal function and convert to BN
+  for (let i = 0; i < distributeArray.length; i++) {
+    let numberWei = ethers.utils.parseEther(distributeArray[i].toString());
+    let BN = BigNumber.from(numberWei);
+    distributeArrayBN.push(BN);
+  }
+  return distributeArrayBN;
```

Comment:
So at first I was about to suggest somehow combining the body of `distribute` with the loop on `48`, but then I noticed your frankly clever use of recursion in `distribute`, which is why a function is necessary there, so that definitely makes sense.

If I might suggest though -- recursive logic is famously hard to understand in a quick read, and I had to actually write stuff out to try to figure out *exactly* how `value` is being distributed. 

On the other hand, recursion can almost always be replaced with easier to understand loops. So in this case, something like:

```
const distributeArrayBN: BigNumber[] = [];
let valueLeft = value;
for (let i=0; i<length; i++) {
  const portion = Math.random() * valueLeft;
  let numberWei = ethers.utils.parseEther(portion.toString());
  let portionBN = BigNumber.from(numberWei);
  distributeArrayBN.push(portionBN);
  valueLeft -= portion;
}
```

If I haven't missed anything and the logic is sound, this should effectively reduce the entire `distributeBN` function to just that one loop.

Reply:
Kelvin's solution is much cleaner. I would suggest moving forward with it.

Reply:
Tried implementing this solution, but the valueLeft number becomes rapidly too small and the loop goes in underflow, 
throwing the error "fractional component exceeds decimals". I tried multiplying portion by 10^n and divide by the same amount later, as well as use Number(amount).toFixed(5) but the issue persists.

A pattern I noticed in the more complex version is that it inserts BN 0 every so often in the array (not in a declining fashion like this updated versione), ie it doesn't go too deep in the decimal operation.

Reply:
Two ideas come to mind, that you could try :

- The initial value is probably too small to begin with. It might not make too much sense distributing 100 units over 50+, esp when `random` is likely to average 0.5 in the first few steps. Try using a much larger initial value?

- If I understood your observation on the pattern seen in the complex version, it sounds like some people in the rewards array don't get assigned any rewards. If that's correct we could try to replicate that behaviour by checking if `valueLeft` at any iteration is too small to further divide up (less than some threshold). If so, assign all of it to the current reward and end the loop.


I am interested in why this pattern exists in the first place though. A happy accident?

Reply:
I tried implementing it with a larger initial number and reducing the length of random distribution, and it works.

This solution, however, doesn't seem too flexible as adding zeros to the remaining rewards that are too small make it look not too random to me as it effectively reduces the length of the random array we are trying to create.

Reply:
We decided to stick with the original version as it accommodates smaller input numbers as well as the output is more round (less decimals)